File: ios/Sources/MqttQuicPlugin/MQTT/MQTT5ReasonCodes.swift
//
// MQTT5ReasonCodes.swift
// MqttQuicPlugin
//
// MQTT 5.0 Reason Codes. Matches MQTTD mqttd/reason_codes.py.
// Uses a struct so multiple names can share the same byte value (e.g. 0x00).
//

import Foundation

public struct MQTT5ReasonCode: Equatable {
    public let rawValue: UInt8
    public init(rawValue: UInt8) { self.rawValue = rawValue }

    // Success (0x00 used in multiple packet types)
    public static let success = MQTT5ReasonCode(rawValue: 0x00)
    public static let normalDisconnection = MQTT5ReasonCode(rawValue: 0x00)
    public static let disconnectWithWillMessage = MQTT5ReasonCode(rawValue: 0x04)

    // CONNACK Reason Codes
    public static let unspecifiedError = MQTT5ReasonCode(rawValue: 0x80)
    public static let malformedPacket = MQTT5ReasonCode(rawValue: 0x81)
    public static let protocolError = MQTT5ReasonCode(rawValue: 0x82)
    public static let implementationSpecificError = MQTT5ReasonCode(rawValue: 0x83)
    public static let unsupportedProtocolVersion = MQTT5ReasonCode(rawValue: 0x84)
    public static let clientIdentifierNotValid = MQTT5ReasonCode(rawValue: 0x85)
    public static let badUserNameOrPassword = MQTT5ReasonCode(rawValue: 0x86)
    public static let notAuthorized = MQTT5ReasonCode(rawValue: 0x87)
    public static let serverUnavailable = MQTT5ReasonCode(rawValue: 0x88)
    public static let serverBusy = MQTT5ReasonCode(rawValue: 0x89)
    public static let banned = MQTT5ReasonCode(rawValue: 0x8A)
    public static let badAuthenticationMethod = MQTT5ReasonCode(rawValue: 0x8C)
    public static let topicNameInvalid = MQTT5ReasonCode(rawValue: 0x90)
    public static let packetTooLarge = MQTT5ReasonCode(rawValue: 0x95)
    public static let quotaExceeded = MQTT5ReasonCode(rawValue: 0x97)
    public static let payloadFormatInvalid = MQTT5ReasonCode(rawValue: 0x99)
    public static let retainNotSupported = MQTT5ReasonCode(rawValue: 0x9A)
    public static let qosNotSupported = MQTT5ReasonCode(rawValue: 0x9B)
    public static let useAnotherServer = MQTT5ReasonCode(rawValue: 0x9C)
    public static let serverMoved = MQTT5ReasonCode(rawValue: 0x9D)
    public static let connectionRateExceeded = MQTT5ReasonCode(rawValue: 0x9F)

    // PUBACK, PUBREC, PUBREL, PUBCOMP Reason Codes
    public static let noMatchingSubscribers = MQTT5ReasonCode(rawValue: 0x10)
    public static let unspecifiedErrorPub = MQTT5ReasonCode(rawValue: 0x80)
    public static let implementationSpecificErrorPub = MQTT5ReasonCode(rawValue: 0x83)
    public static let notAuthorizedPub = MQTT5ReasonCode(rawValue: 0x87)
    public static let topicNameInvalidPub = MQTT5ReasonCode(rawValue: 0x90)
    public static let packetIdentifierInUse = MQTT5ReasonCode(rawValue: 0x91)
    public static let quotaExceededPub = MQTT5ReasonCode(rawValue: 0x97)
    public static let payloadFormatInvalidPub = MQTT5ReasonCode(rawValue: 0x99)

    // SUBACK Reason Codes
    public static let grantedQoS0 = MQTT5ReasonCode(rawValue: 0x00)
    public static let grantedQoS1 = MQTT5ReasonCode(rawValue: 0x01)
    public static let grantedQoS2 = MQTT5ReasonCode(rawValue: 0x02)
    public static let unspecifiedErrorSub = MQTT5ReasonCode(rawValue: 0x80)
    public static let implementationSpecificErrorSub = MQTT5ReasonCode(rawValue: 0x83)
    public static let notAuthorizedSub = MQTT5ReasonCode(rawValue: 0x87)
    public static let topicFilterInvalid = MQTT5ReasonCode(rawValue: 0x8F)
    public static let packetIdentifierInUseSub = MQTT5ReasonCode(rawValue: 0x91)
    public static let quotaExceededSub = MQTT5ReasonCode(rawValue: 0x97)
    public static let sharedSubscriptionsNotSupported = MQTT5ReasonCode(rawValue: 0x9E)
    public static let subscriptionIdentifiersNotSupported = MQTT5ReasonCode(rawValue: 0xA1)
    public static let wildcardSubscriptionsNotSupported = MQTT5ReasonCode(rawValue: 0xA2)

    // UNSUBACK Reason Codes
    public static let successUnsub = MQTT5ReasonCode(rawValue: 0x00)
    public static let noSubscriptionExisted = MQTT5ReasonCode(rawValue: 0x11)
    public static let unspecifiedErrorUnsub = MQTT5ReasonCode(rawValue: 0x80)
    public static let implementationSpecificErrorUnsub = MQTT5ReasonCode(rawValue: 0x83)
    public static let notAuthorizedUnsub = MQTT5ReasonCode(rawValue: 0x87)
    public static let topicFilterInvalidUnsub = MQTT5ReasonCode(rawValue: 0x8F)
    public static let packetIdentifierInUseUnsub = MQTT5ReasonCode(rawValue: 0x91)

    // DISCONNECT Reason Codes
    public static let normalDisconnectionDisc = MQTT5ReasonCode(rawValue: 0x00)
    public static let disconnectWithWillMessageDisc = MQTT5ReasonCode(rawValue: 0x04)
    public static let unspecifiedErrorDisc = MQTT5ReasonCode(rawValue: 0x80)
    public static let malformedPacketDisc = MQTT5ReasonCode(rawValue: 0x81)
    public static let protocolErrorDisc = MQTT5ReasonCode(rawValue: 0x82)
    public static let implementationSpecificErrorDisc = MQTT5ReasonCode(rawValue: 0x83)
    public static let notAuthorizedDisc = MQTT5ReasonCode(rawValue: 0x87)
    public static let serverBusyDisc = MQTT5ReasonCode(rawValue: 0x89)
    public static let serverShuttingDown = MQTT5ReasonCode(rawValue: 0x8B)
    public static let badAuthenticationMethodDisc = MQTT5ReasonCode(rawValue: 0x8C)
    public static let keepAliveTimeout = MQTT5ReasonCode(rawValue: 0x8D)
    public static let sessionTakenOver = MQTT5ReasonCode(rawValue: 0x8E)
    public static let topicFilterInvalidDisc = MQTT5ReasonCode(rawValue: 0x8F)
    public static let topicNameInvalidDisc = MQTT5ReasonCode(rawValue: 0x90)
    public static let receiveMaximumExceeded = MQTT5ReasonCode(rawValue: 0x93)
    public static let topicAliasInvalid = MQTT5ReasonCode(rawValue: 0x94)
    public static let packetTooLargeDisc = MQTT5ReasonCode(rawValue: 0x95)
    public static let messageRateTooHigh = MQTT5ReasonCode(rawValue: 0x96)
    public static let quotaExceededDisc = MQTT5ReasonCode(rawValue: 0x97)
    public static let administrativeAction = MQTT5ReasonCode(rawValue: 0x98)
    public static let payloadFormatInvalidDisc = MQTT5ReasonCode(rawValue: 0x99)
    public static let retainNotSupportedDisc = MQTT5ReasonCode(rawValue: 0x9A)
    public static let qosNotSupportedDisc = MQTT5ReasonCode(rawValue: 0x9B)
    public static let useAnotherServerDisc = MQTT5ReasonCode(rawValue: 0x9C)
    public static let serverMovedDisc = MQTT5ReasonCode(rawValue: 0x9D)
    public static let sharedSubscriptionsNotSupportedDisc = MQTT5ReasonCode(rawValue: 0x9E)
    public static let connectionRateExceededDisc = MQTT5ReasonCode(rawValue: 0x9F)
    public static let maximumConnectTime = MQTT5ReasonCode(rawValue: 0xA0)
    public static let subscriptionIdentifiersNotSupportedDisc = MQTT5ReasonCode(rawValue: 0xA1)
    public static let wildcardSubscriptionsNotSupportedDisc = MQTT5ReasonCode(rawValue: 0xA2)
}

// Compatibility mapping for MQTT 3.1.1 return codes
public let MQTT3_TO_MQTT5_REASON_CODE: [UInt8: MQTT5ReasonCode] = [
    0x00: .success,
    0x01: .unsupportedProtocolVersion,
    0x02: .clientIdentifierNotValid,
    0x03: .serverUnavailable,
    0x04: .badUserNameOrPassword,
    0x05: .notAuthorized,
]

File: android/src/main/kotlin/ai/annadata/mqttquic/mqtt/MQTT5ReasonCodes.kt
package ai.annadata.mqttquic.mqtt

/**
 * MQTT 5.0 Reason Codes. Matches MQTTD mqttd/reason_codes.py.
 * MQTT 5.0 reuses byte values across packet types; we keep one constant per unique byte.
 * Context-specific aliases are provided as properties for semantic clarity.
 */
object MQTT5ReasonCode {
    // 0x00 - Success (CONNACK), Granted QoS 0 (SUBACK), Success (UNSUBACK), Normal disconnection (DISCONNECT)
    const val SUCCESS: Int = 0x00
    // 0x01 - Granted QoS 1 (SUBACK)
    const val GRANTED_QOS_1: Int = 0x01
    // 0x02 - Granted QoS 2 (SUBACK)
    const val GRANTED_QOS_2: Int = 0x02
    // 0x04 - Disconnect with will message
    const val DISCONNECT_WITH_WILL_MESSAGE: Int = 0x04
    // 0x10 - No matching subscribers (PUBACK/PUBREC/PUBREL/PUBCOMP)
    const val NO_MATCHING_SUBSCRIBERS: Int = 0x10
    // 0x11 - No subscription existed (UNSUBACK)
    const val NO_SUBSCRIPTION_EXISTED: Int = 0x11
    
    // CONNACK / generic errors (0x80+)
    const val UNSPECIFIED_ERROR: Int = 0x80
    const val MALFORMED_PACKET: Int = 0x81
    const val PROTOCOL_ERROR: Int = 0x82
    const val IMPLEMENTATION_SPECIFIC_ERROR: Int = 0x83
    const val UNSUPPORTED_PROTOCOL_VERSION: Int = 0x84
    const val CLIENT_IDENTIFIER_NOT_VALID: Int = 0x85
    const val BAD_USER_NAME_OR_PASSWORD: Int = 0x86
    const val NOT_AUTHORIZED: Int = 0x87
    const val SERVER_UNAVAILABLE: Int = 0x88
    const val SERVER_BUSY: Int = 0x89
    const val BANNED: Int = 0x8A
    const val SERVER_SHUTTING_DOWN: Int = 0x8B
    const val BAD_AUTHENTICATION_METHOD: Int = 0x8C
    const val KEEP_ALIVE_TIMEOUT: Int = 0x8D
    const val SESSION_TAKEN_OVER: Int = 0x8E
    const val TOPIC_FILTER_INVALID: Int = 0x8F
    const val TOPIC_NAME_INVALID: Int = 0x90
    const val PACKET_IDENTIFIER_IN_USE: Int = 0x91
    const val RECEIVE_MAXIMUM_EXCEEDED: Int = 0x93
    const val TOPIC_ALIAS_INVALID: Int = 0x94
    const val PACKET_TOO_LARGE: Int = 0x95
    const val MESSAGE_RATE_TOO_HIGH: Int = 0x96
    const val QUOTA_EXCEEDED: Int = 0x97
    const val ADMINISTRATIVE_ACTION: Int = 0x98
    const val PAYLOAD_FORMAT_INVALID: Int = 0x99
    const val RETAIN_NOT_SUPPORTED: Int = 0x9A
    const val QOS_NOT_SUPPORTED: Int = 0x9B
    const val USE_ANOTHER_SERVER: Int = 0x9C
    const val SERVER_MOVED: Int = 0x9D
    const val SHARED_SUBSCRIPTIONS_NOT_SUPPORTED: Int = 0x9E
    const val CONNECTION_RATE_EXCEEDED: Int = 0x9F
    const val MAXIMUM_CONNECT_TIME: Int = 0xA0
    const val SUBSCRIPTION_IDENTIFIERS_NOT_SUPPORTED: Int = 0xA1
    const val WILDCARD_SUBSCRIPTIONS_NOT_SUPPORTED: Int = 0xA2
    
    // Context-specific aliases (all map to the same underlying values)
    // SUBACK aliases
    const val GRANTED_QOS_0: Int = SUCCESS  // 0x00
    
    // UNSUBACK aliases
    const val SUCCESS_UNSUB: Int = SUCCESS  // 0x00
    
    // DISCONNECT aliases
    const val NORMAL_DISCONNECTION: Int = SUCCESS  // 0x00
    const val NORMAL_DISCONNECTION_DISC: Int = SUCCESS  // 0x00
    const val DISCONNECT_WITH_WILL_MESSAGE_DISC: Int = DISCONNECT_WITH_WILL_MESSAGE  // 0x04
    const val UNSPECIFIED_ERROR_DISC: Int = UNSPECIFIED_ERROR  // 0x80
    const val MALFORMED_PACKET_DISC: Int = MALFORMED_PACKET  // 0x81
    const val PROTOCOL_ERROR_DISC: Int = PROTOCOL_ERROR  // 0x82
    const val IMPLEMENTATION_SPECIFIC_ERROR_DISC: Int = IMPLEMENTATION_SPECIFIC_ERROR  // 0x83
    const val NOT_AUTHORIZED_DISC: Int = NOT_AUTHORIZED  // 0x87
    const val SERVER_BUSY_DISC: Int = SERVER_BUSY  // 0x89
    const val BAD_AUTHENTICATION_METHOD_DISC: Int = BAD_AUTHENTICATION_METHOD  // 0x8C
    const val TOPIC_FILTER_INVALID_DISC: Int = TOPIC_FILTER_INVALID  // 0x8F
    const val TOPIC_NAME_INVALID_DISC: Int = TOPIC_NAME_INVALID  // 0x90
    const val PACKET_TOO_LARGE_DISC: Int = PACKET_TOO_LARGE  // 0x95
    const val QUOTA_EXCEEDED_DISC: Int = QUOTA_EXCEEDED  // 0x97
    const val PAYLOAD_FORMAT_INVALID_DISC: Int = PAYLOAD_FORMAT_INVALID  // 0x99
    const val RETAIN_NOT_SUPPORTED_DISC: Int = RETAIN_NOT_SUPPORTED  // 0x9A
    const val QOS_NOT_SUPPORTED_DISC: Int = QOS_NOT_SUPPORTED  // 0x9B
    const val USE_ANOTHER_SERVER_DISC: Int = USE_ANOTHER_SERVER  // 0x9C
    const val SERVER_MOVED_DISC: Int = SERVER_MOVED  // 0x9D
    const val SHARED_SUBSCRIPTIONS_NOT_SUPPORTED_DISC: Int = SHARED_SUBSCRIPTIONS_NOT_SUPPORTED  // 0x9E
    const val CONNECTION_RATE_EXCEEDED_DISC: Int = CONNECTION_RATE_EXCEEDED  // 0x9F
    const val SUBSCRIPTION_IDENTIFIERS_NOT_SUPPORTED_DISC: Int = SUBSCRIPTION_IDENTIFIERS_NOT_SUPPORTED  // 0xA1
    const val WILDCARD_SUBSCRIPTIONS_NOT_SUPPORTED_DISC: Int = WILDCARD_SUBSCRIPTIONS_NOT_SUPPORTED  // 0xA2
    
    // PUBACK/PUBREC/PUBREL/PUBCOMP aliases
    const val UNSPECIFIED_ERROR_PUB: Int = UNSPECIFIED_ERROR  // 0x80
    const val IMPLEMENTATION_SPECIFIC_ERROR_PUB: Int = IMPLEMENTATION_SPECIFIC_ERROR  // 0x83
    const val NOT_AUTHORIZED_PUB: Int = NOT_AUTHORIZED  // 0x87
    const val TOPIC_NAME_INVALID_PUB: Int = TOPIC_NAME_INVALID  // 0x90
    const val QUOTA_EXCEEDED_PUB: Int = QUOTA_EXCEEDED  // 0x97
    const val PAYLOAD_FORMAT_INVALID_PUB: Int = PAYLOAD_FORMAT_INVALID  // 0x99
    
    // SUBACK aliases
    const val UNSPECIFIED_ERROR_SUB: Int = UNSPECIFIED_ERROR  // 0x80
    const val IMPLEMENTATION_SPECIFIC_ERROR_SUB: Int = IMPLEMENTATION_SPECIFIC_ERROR  // 0x83
    const val NOT_AUTHORIZED_SUB: Int = NOT_AUTHORIZED  // 0x87
    const val PACKET_IDENTIFIER_IN_USE_SUB: Int = PACKET_IDENTIFIER_IN_USE  // 0x91
    const val QUOTA_EXCEEDED_SUB: Int = QUOTA_EXCEEDED  // 0x97
    
    // UNSUBACK aliases
    const val UNSPECIFIED_ERROR_UNSUB: Int = UNSPECIFIED_ERROR  // 0x80
    const val IMPLEMENTATION_SPECIFIC_ERROR_UNSUB: Int = IMPLEMENTATION_SPECIFIC_ERROR  // 0x83
    const val NOT_AUTHORIZED_UNSUB: Int = NOT_AUTHORIZED  // 0x87
    const val TOPIC_FILTER_INVALID_UNSUB: Int = TOPIC_FILTER_INVALID  // 0x8F
    const val PACKET_IDENTIFIER_IN_USE_UNSUB: Int = PACKET_IDENTIFIER_IN_USE  // 0x91
}

// Compatibility mapping for MQTT 3.1.1 return codes
val MQTT3_TO_MQTT5_REASON_CODE = mapOf(
    0x00 to MQTT5ReasonCode.SUCCESS,
    0x01 to MQTT5ReasonCode.UNSUPPORTED_PROTOCOL_VERSION,
    0x02 to MQTT5ReasonCode.CLIENT_IDENTIFIER_NOT_VALID,
    0x03 to MQTT5ReasonCode.SERVER_UNAVAILABLE,
    0x04 to MQTT5ReasonCode.BAD_USER_NAME_OR_PASSWORD,
    0x05 to MQTT5ReasonCode.NOT_AUTHORIZED,
)
